src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let clonedState = Object.assign({}, state);
+
+  const actionHandlers = {
+    addProperties: (action) => {
+      clonedState = Object.assign({}, clonedState, action.extraData);
+      result.push(clonedState);
+    },
+    removeProperties: (action) => {
+      const keysToRemove = action.keysToRemove;
+      const nextState = Object.assign({}, clonedState);
+
+      for (const key of keysToRemove) {
+        if (nextState.hasOwnProperty(key)) {
+          delete nextState[key];
+        }
+      }
+
+      clonedState = nextState;
+      result.push(clonedState);
+    },
+    clear: () => {
+      clonedState = {};
+      result.push(clonedState);
+    },
```

Comment:
no need to make a push in each handler. You can do it once in the cycle at the end of each iteration